### PR TITLE
Issue/10/function chainer

### PR DESCRIPTION
**Pull request by ellesmith88**
Adapted fixer process to support multiple steps.

### CODE REVIEW

**Comment by agstephens** [comment]
@ellesmith88 : It would be great to include a couple of unit tests to make sure this is working properly. 

Also, this line looks like it probably needs changing:
```
        for f in self.funcs[0]:
```

Should it be as follows...to make sure all functions are being chained together?

```
        for f in self.funcs:
```

**Comment by ellesmith88** [comment]
@agstephens Added tests and fixed the FuncChainer to `for f in self.funcs:`

**Comment by agstephens** [comment]
Hi @ellesmith88: would you look at making the tests for the chaining slightly more specific in terms of testing the final data file/object to assert that the correct processing has happened? This could be done with something like:

 - preprocessor 1: 
    - `tests.dummy_data_utils.double_array` - which multiples the array x 2
 - preprocessor 2:
    - `tests.dummy_data_utils.set_long_name_silly` - which sets the variable `long_name` attribute to "silly"
 - postprocessor 1:
    - `tests.dummy_data_utils.update_attrs` - updates variable attributes, with keyword args:
      - {"test_attr_1": "hello world", "units": "m/s"}
 - postprocessor 2:
    - `tests.dummy_data_utils.add_number` - adds the number provided as an argument to the array:
      - [15.5]

Then your assertions in the test compare:
 1. Read the data directly in using `xarray.open_mfdataset()`
    - apply those fixes above in simple xarray code
       - e.g. `da2 = da * 2`, `da2.attrs['long_name'] = 'silly'`, etc, etc.
 2. Read the same data using `daops.utils.open_dataset()`:
    - which should have applied the fixes itself.
 3. Compare those characteristics of the two datasets that have been modified in the processors.
    - use an `assert()` for each

This kind of approach will help us believe that our system is working properly and our xarray confidence is also being built up. 👍

**Comment by ellesmith88** [comment]
@agstephens I've updated the tests. For things like long_name and standard_name you need to access this using the var_id, which needs to be passed in.
- I'm not sure I 'm doing this in the best way.
- This would mean that things changed which need the var_id need to be done via a  post_processor?

**Comment by agstephens** [comment]
@ellesmith88, good point. So there is a need to _identify_ the main variable in the code. You could write a function called `get_main_variable` that finds the variable from something like:

```
var_id = sorted([(data.size, var_id) for var_id, data in ds.variables.items()])[-1][1]
```

You might want to unpack this see what it does, but I'm assuming the variable with the largest size is the main variable. That might be a bad assumption because you might get some grids coordinate variables that are the same size as the main variable. 

So you might want to add in something that checks that the chosen variable is not in the `dims` list of any other variable...

```
>>> [(var_id, data.dims) for var_id, data in ds.variables.items()]
[('time', ('time',)), ('time_bnds', ('time', 'bnds')), ('lat', ('lat',)), ('lat_bnds', ('lat', 'bnds')), ('lon', ('lon',)), ('lon_bnds', ('lon', 'bnds')), ('rh', ('time', 'lat', 'lon'))]
```

To summarise:
 - get the variables that have the biggest arrays
 - then compare their `dims` list and choose the one that is not in the list for any other.
 - if len(result) != 1: raise an Exception saying you could not determine the _main_ variable.

**Comment by agstephens** [comment]
@ellesmith88 , thanks for putting this together. It looks comprehensive. One minor adjustment before we merge. I note that you have defined `double_lats`. I reckon this should be `double_array` - i.e. doubling the main variable values rather than the latitude values. Apart from that it looks great. Thanks

**Comment by ellesmith88** [comment]
@agstephens Changed that now 😄
